experimental/Geometry/Multiindices.jl

```diff
@@ -0,0 +1,221 @@
+module Multiindices
+
+export OrderedMultiindex
+export linear_index
+export index_signature
+export HomogMultiindex
+export power
+export getindex
+
+
+abstract type Multiindex end
+
+mutable struct OrderedMultiindex <: Multiindex 
+  n::Int
+  p::Int
+  index::Vector{Int}
+  function OrderedMultiindex(n::Int, p::Int)
+    v = Vector{Int}(undef,p) 
+    for i in 1:p
+      v[i] = i
+    end
+    #@show n, p, v
+    return new( n, p, v )
+  end
+end
+
+function Base.show( io::Base.TTY, i::OrderedMultiindex ) 
+  outstr= "0"
+  for k in 1:length(i.index) 
+    outstr = outstr * " < " * string(i.index[k]) 
+  end
+  outstr = outstr * " <= " * string( i.n ) * "\n"
+  #Base.show( io, "n = " * string(i.n) * "; p = " * string(i.p) * "; index = " * string( i.index ) )
+  Base.print( io, outstr )
+end
+
+function Base.iterate( i::OrderedMultiindex ) 
+  for k in 1:i.p
+    i.index[k] = k
```

Comment:
Really? From what I was reading about implementing the iteration interface, this was the way to go. Also, you don't iterate over a range of integers without changing the value of your variable, don't you?

Reply:
No, of course an iterator should not modify the object it "iterates over". That's what the "state" variable is for -- it stores the "state" of the iteration.

To take your example of "iterating over a range of integers": what you are doing is akin to *modifying the range of integers* over which you iterate.

Reply:
(It is not forbidden to have stateful iterators in julia.)

Reply:
True, but it wasn't clear to me that `OrderedMultiindex` is an iterator type? If so, then it should be renamed to make that clear.

Reply:
> No, of course an iterator should not modify the object it "iterates over". That's what the "state" variable is for -- it stores the "state" of the iteration.
> 
> To take your example of "iterating over a range of integers": what you are doing is akin to _modifying the range of integers_ over which you iterate.

I see your point. However, if you look at the design of the multiindex (which is admittedly very poorly commented), then you see that the range is fixed by the values of n and p. The i (or a) accounts for the state variable. And this is the only thing that is being touched. 

However, I will review this regarding the question whether the state variable can be used for this.

Reply:
I think it fine that this is the iterator type itself and storing the state inside can be done. But then, the first return value should not be the iterator itself, but maybe a `Vector` representing the ordered index itself? Because this is what the `a` is in `for a in i`.

Reply:
But inside the loop one will want to do something with the index a. And as I said before, the set of increasing integers in a are not worth much unless one also provides the context n and p. So I preferred to have the whole object returned. Which has the further advantage that one can use the routines developed for handling ordered Multiindices and one does not have to do things by hand on a vector of integers. 

But I feel this is for a longer discussion that we probably don't want to have here, typing everything.

Reply:
I see, sure! Then there should be two distinct types. The iterator objects as well as the things that are returned, which are the multiindices themselves. Mixing both will probably lead to problems down the road, since this is not how iterators are supposed to be implemented in julia.